salt/utils/data.py

```diff
@@ -765,6 +765,14 @@ def traverse_dict_and_list(data, key, default=None, delimiter=DEFAULT_TARGET_DEL
                     ptr = ptr[idx]
                 except IndexError:
                     return default
+        elif isinstance(ptr, dict):
+            try:
+                for key, value in ptr.items():
+                    if not isinstance(key, str):
+                        ptr[str(key)] = ptr.pop(key)
+                ptr = ptr[each]
+            except KeyError:
+                return default
```

Comment:
You should not change the dict while iterating on it. I got it working by mapping the dictionary keys to they string representations, however this could have a performance impact.

```suggestion
            try:
                subkeys = {str(subkey): subkey for subkey in ptr.keys()}
                subkey = subkeys[each]
                ptr = ptr[subkey]
            except KeyError:
                return default
```

In order to avoid adding loops on potentially long dicts, you can do it in an uglier and less pythonic way like this:

``` suggestion
            found_key = False
            for subkey in ptr.keys():
                if str(subkey) == each:
                    found_key = True
                    ptr = ptr[subkey]
                    break
            if not found_key:
                return default
```

Here I would prefer the performance-safe option. @dwoz any opinion?

Also please do not use  `key` a a local variable since it will shadow the parameter of the same name.

Reply:
Shall I do this in the below way? I think it is compatible with all versions. @cbosdo @dwoz 
```
elif isinstance(ptr, dict):
            try:
                ptr={str(k):v for k, v in ptr.items()}
                ptr = ptr[each]
            except KeyError:
                return default
```

Reply:
This would be a bit faster in case the `each` value is in dict keys.
```python
        elif isinstance(ptr, dict):
            try:
                ptr = ptr[each]  # First try the fastest way
            except KeyError:
                try:
                    ptr={str(k):v for k, v in ptr.items() if isinstance(k, six.string_types)}
                    ptr = ptr[each]
                except KeyError:
                    return default
```

Reply:
yeah, okay.

Reply:
using this ```if isinstance(k, six.string_types)```condition will allow only string type key inside the dict. The key issue occur in nested dict like ```ptr= {"foo": {1234: "bar"}}``` but ```each``` value of type str '1234' will not reach ```bar```. The below one works,

```python
elif isinstance(ptr, dict):
            try:
                ptr = ptr[each]  # First try the fastest way
            except KeyError:
                try:
                    ptr={str(k):v for k, v in ptr.items()}
                    ptr = ptr[each]
                except KeyError:
                    return default
```

Reply:
I have submitted an alternate fix which doesn't require type checking nor iteration.

https://github.com/saltstack/salt/pull/56748

Reply:
> I have submitted an alternate fix that doesn't require type checking nor iteration.
> 
> #56748

Thanks for knowing me this function. I will find some other interesting issues and do my contribution to this repo. Shall I close this PR now?

Reply:
Yes, please, and thanks for taking the time to try to contribute.